File: go/groupSam.go
package main

import (
    "os"
    "fmt"
    //"regexp"
    "bufio"
    "strings"
)

func main() {
    if len(os.Args) != 2 {
		fmt.Println("### Criado por Tiago Minuzzi ###","\n")
		fmt.Println("Uso: ./groupSam nome_do_arquivo.sam > saida.tsv")
		os.Exit(1)
	}

    arquivo, err := os.Open(os.Args[1])
    if err != nil {
        fmt.Printf("Erro: %s ao abrir o arquivo %s", err, arquivo)
    }

    defer arquivo.Close()

    groupedSam := createSamMap(arquivo)

    var resultado strings.Builder

    for i, j := range groupedSam {
        resultado.WriteString(i)
        resultado.WriteString("\t")
        resultado.WriteString(strings.Join(j, "\t"))
        resultado.WriteString("\n")
    }

    fmt.Println(resultado.String())
}

func createSamMap(samFile *os.File) map[string][]string {
    reader := bufio.NewScanner(samFile)
    
    samMap := make(map[string][]string)

    for reader.Scan() {
        linha := reader.Text()
        linha = strings.TrimSpace(linha)

        if !strings.HasPrefix(linha, "@"){
            splitted := strings.Split(linha, "\t")
            idx,_,cgr,_ := splitted[0], splitted[1:5], splitted[5], splitted[6:]
            samMap[idx] = append(samMap[idx], cgr)
        }
    }
    return samMap
}

File: rust/groupSam.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead};
use std::env;

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();

    // Verifique se o nome do arquivo foi fornecido
    if args.len() < 2 {
        eprintln!("Uso: {} <nome_do_arquivo>", args[0]);
        std::process::exit(1);
    }

    let file_path = &args[1];

    let file = File::open(&file_path)?;
    let file = io::BufReader::new(file);

    let mut data: HashMap<String, Vec<String>> = HashMap::new();

    for linha in file.lines() {
        let linha: String = linha.expect("Erro ao ler a linha.");
        if !linha.starts_with('@') {
            let linha: Vec<&str> = linha.split("\t").collect();
            let idx = linha[0].to_string();
            let cgr = linha[5].to_string();

            data.entry(idx).or_insert_with(Vec::new).push(cgr);
        }
    }

    for (k, v) in data {
        println!("{}\t{}",k, v.join("\t"));
    }
    Ok(())
}
